File: edge/ios/Sources/AgatEdgeCore/CoordinatorClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public enum CoordinatorClientError: Error, Equatable {
    case invalidResponse
    case responseTooLarge
    case http(Int, String)
    case unsupportedLease
}

private final class NoRedirectDelegate: NSObject, URLSessionTaskDelegate, @unchecked Sendable {
    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        completionHandler(nil)
    }
}

public actor CoordinatorClient {
    private let configuration: EdgeConfiguration
    private let session: URLSession
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    public init(configuration: EdgeConfiguration, session: URLSession? = nil) {
        self.configuration = configuration
        if let session {
            self.session = session
        } else {
            let settings = URLSessionConfiguration.ephemeral
            settings.timeoutIntervalForRequest = 15
            settings.timeoutIntervalForResource = 75
            settings.httpCookieStorage = nil
            settings.urlCredentialStorage = nil
            self.session = URLSession(configuration: settings, delegate: NoRedirectDelegate(), delegateQueue: nil)
        }
        self.encoder = JSONEncoder()
        self.decoder = JSONDecoder()
        self.encoder.dateEncodingStrategy = .iso8601
        self.decoder.dateDecodingStrategy = .iso8601
    }

    public func requestChallenge(bootstrapToken: String, applicationId: String) async throws -> EnrollmentChallenge {
        struct Body: Encodable {
            let enrollmentToken: String
            let name: String
            let platform: String
            let applicationId: String
        }
        let data = try await request(
            path: "/api/v1/edge/enrollment/challenges",
            method: "POST",
            body: Body(enrollmentToken: bootstrapToken, name: configuration.nodeName, platform: "ios", applicationId: applicationId),
            expectedStatus: 201
        )
        return try decoder.decode(EnrollmentChallenge.self, from: data)
    }

    public func enroll(
        challenge: EnrollmentChallenge,
        evidence: AttestationEvidence,
        modelSizeBytes: Int64,
        memoryMb: Int
    ) async throws -> EnrollmentResponse {
        struct ModelProfile: Encodable {
            let name: String
            let provider = "coreml"
            let sizeBytes: Int64
            let quantization = "coreml"
            let capabilities = ["text"]
        }
        struct Body: Encodable {
            let challengeId: String
            let challenge: String
            let name: String
            let platform = "ios"
            let architecture = "arm64"
            let models: [String]
            let modelProfiles: [ModelProfile]
            let embeddingModels: [String] = []
            let maxConcurrency = 1
            let memoryMb: Int
            let gpu = "Core ML / Metal"
            let labels: [String: String]
            let attestation: AttestationEvidence
        }
        let body = Body(
            challengeId: challenge.id,
            challenge: challenge.challenge,
            name: configuration.nodeName,
            models: [configuration.modelName],
            modelProfiles: [ModelProfile(name: configuration.modelName, sizeBytes: modelSizeBytes)],
            memoryMb: memoryMb,
            labels: ["runtime": "ios-app", "inference": "coreml+metal", "background": "os-scheduled"],
            attestation: evidence
        )
        let data = try await request(path: "/api/v1/edge/enroll", method: "POST", body: body, expectedStatus: 201)
        return try decoder.decode(EnrollmentResponse.self, from: data)
    }

    public func heartbeat(token: String, batteryPercent: Double?, onBattery: Bool) async throws -> ControlCommand {
        struct Metrics: Encodable { let batteryPercent: Double?; let onBattery: Bool }
        struct Profile: Encodable {
            let name: String
            let provider = "coreml"
            let quantization = "coreml"
            let capabilities = ["text"]
        }
        struct Capabilities: Encodable {
            let models: [String]
            let modelProfiles: [Profile]
            let embeddingModels: [String] = []
            let agentRuntimes = ["single"]
            let agentRuntimeProfiles = ["tool_loop_v1"]
            let maxConcurrency = 1
            let labels = ["runtime": "ios-app", "inference": "coreml+metal", "background": "os-scheduled"]
        }
        struct Body: Encodable { let metrics: Metrics; let capabilities: Capabilities }
        let body = Body(
            metrics: Metrics(batteryPercent: batteryPercent, onBattery: onBattery),
            capabilities: Capabilities(models: [configuration.modelName], modelProfiles: [Profile(name: configuration.modelName)])
        )
        let data = try await request(path: "/api/v1/workers/heartbeat", method: "POST", body: body, bearer: token, expectedStatus: 200)
        return try decoder.decode(ControlCommand.self, from: data)
    }

    public func lease(token: String) async throws -> EdgeLease? {
        struct Body: Encodable { let workerVersion = "ios/1.6.0" }
        let result = try await rawRequest(path: "/api/v1/workers/lease", method: "POST", body: Body(), bearer: token)
        if result.status == 204 { return nil }
        guard result.status == 200 else { throw httpError(status: result.status, data: result.data) }
        let lease = try decoder.decode(EdgeLease.self, from: result.data)
        guard lease.activity == nil, lease.mcpTools.isEmpty, lease.agent.runtime == "single" else {
            throw CoordinatorClientError.unsupportedLease
        }
        return lease
    }

    public func renew(token: String, leaseId: String) async throws {
        try await emptyRequest(path: "/api/v1/leases/\(pathSegment(leaseId))/renew", body: Empty(), bearer: token, expectedStatus: 204)
    }

    public func complete(token: String, leaseId: String, output: String, durationMs: Int) async throws {
        struct Metrics: Encodable {
            let durationMs: Int
            let modelCalls = 1
            let provider = "coreml-metal"
            let model: String
        }
        struct Body: Encodable { let output: String; let metrics: Metrics }
        try await emptyRequest(
            path: "/api/v1/leases/\(pathSegment(leaseId))/complete",
            body: Body(output: output, metrics: Metrics(durationMs: durationMs, model: configuration.modelName)),
            bearer: token,
            expectedStatus: 200
        )
    }

    public func fail(token: String, leaseId: String, error: String) async throws {
        struct Body: Encodable { let error: String }
        let safe = String(error.replacingOccurrences(of: "\n", with: " ").replacingOccurrences(of: "\r", with: " ").prefix(2_000))
        try await emptyRequest(
            path: "/api/v1/leases/\(pathSegment(leaseId))/fail",
            body: Body(error: safe.isEmpty ? "Native inference failed" : safe),
            bearer: token,
            expectedStatus: 200
        )
    }

    public func acknowledgeWipe(token: String, generation: Int, localDataDeleted: Bool) async throws {
        struct Body: Encodable {
            let generation: Int
            let credentialsDeleted = true
            let localDataDeleted: Bool
        }
        try await emptyRequest(
            path: "/api/v1/edge/control/wipe-ack",
            body: Body(generation: generation, localDataDeleted: localDataDeleted),
            bearer: token,
            expectedStatus: 200
        )
    }

    private func emptyRequest<T: Encodable>(path: String, body: T, bearer: String?, expectedStatus: Int) async throws {
        _ = try await request(path: path, method: "POST", body: body, bearer: bearer, expectedStatus: expectedStatus)
    }

    private func request<T: Encodable>(
        path: String,
        method: String,
        body: T,
        bearer: String? = nil,
        expectedStatus: Int
    ) async throws -> Data {
        let result = try await rawRequest(path: path, method: method, body: body, bearer: bearer)
        guard result.status == expectedStatus else { throw httpError(status: result.status, data: result.data) }
        return result.data
    }

    private func rawRequest<T: Encodable>(path: String, method: String, body: T, bearer: String?) async throws -> (status: Int, data: Data) {
        guard let url = URL(string: path, relativeTo: configuration.coordinatorURL)?.absoluteURL else {
            throw EdgeConfigurationError.invalidCoordinatorURL
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.httpBody = try encoder.encode(body)
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("agat-edge-ios/1.6.0", forHTTPHeaderField: "User-Agent")
        if let bearer { request.setValue("Bearer \(bearer)", forHTTPHeaderField: "Authorization") }
        let (data, response) = try await session.data(for: request)
        guard data.count <= 1_048_576 else { throw CoordinatorClientError.responseTooLarge }
        guard let response = response as? HTTPURLResponse else { throw CoordinatorClientError.invalidResponse }
        return (response.statusCode, data)
    }

    private func httpError(status: Int, data: Data) -> CoordinatorClientError {
        let detail = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["error"] as? String ?? ""
        return .http(status, String(detail.prefix(500)))
    }

    private func pathSegment(_ value: String) -> String {
        value.addingPercentEncoding(withAllowedCharacters: .alphanumerics) ?? "invalid"
    }

    private struct Empty: Encodable {}
}

File: edge/android/app/src/main/java/io/agat/edge/CoordinatorClient.kt
package io.agat.edge

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.io.ByteArrayOutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import android.util.Base64

data class EnrollmentChallenge(val id: String, val challenge: String, val expiresAt: String)
data class EnrolledNode(val id: String, val token: String)
data class ControlCommand(val action: String, val generation: Int, val reason: String?)

class CoordinatorClient(private val configuration: EdgeConfiguration) {
    private val baseUrl = configuration.validate().coordinatorUrl

    suspend fun requestChallenge(bootstrapToken: String): EnrollmentChallenge {
        val payload = JSONObject()
            .put("enrollmentToken", bootstrapToken)
            .put("name", configuration.nodeName)
            .put("platform", "android")
            .put("applicationId", BuildConfig.APPLICATION_ID)
        val response = request("POST", "/api/v1/edge/enrollment/challenges", payload)
        requireStatus(response, 201)
        val json = JSONObject(response.body)
        return EnrollmentChallenge(json.getString("id"), json.getString("challenge"), json.getString("expiresAt"))
    }

    suspend fun enroll(
        challenge: EnrollmentChallenge,
        attestationToken: String,
        installationKey: InstallationKey,
        modelSizeBytes: Long,
        memoryMb: Long,
    ): EnrolledNode {
        val capabilities = JSONArray().put("text")
        val modelProfiles = JSONArray().put(
            JSONObject()
                .put("name", configuration.modelName)
                .put("provider", "llama.cpp")
                .put("sizeBytes", modelSizeBytes)
                .put("quantization", "gguf")
                .put("capabilities", capabilities),
        )
        val payload = JSONObject()
            .put("challengeId", challenge.id)
            .put("challenge", challenge.challenge)
            .put("name", configuration.nodeName)
            .put("platform", "android")
            .put("architecture", android.os.Build.SUPPORTED_ABIS.firstOrNull() ?: "unknown")
            .put("models", JSONArray().put(configuration.modelName))
            .put("modelProfiles", modelProfiles)
            .put("embeddingModels", JSONArray())
            .put("maxConcurrency", 1)
            .put("memoryMb", memoryMb)
            .put("gpu", if (configuration.backend == InferenceBackend.VULKAN) "Vulkan" else "CPU")
            .put("labels", JSONObject()
                .put("runtime", "android-service")
                .put("inference", if (configuration.backend == InferenceBackend.VULKAN) "llama.cpp+vulkan" else "llama.cpp+cpu")
                .put("nnapi", NativeLlama.nnapiCapabilityLabel())
                .put("installationHardwareBacked", installationKey.hardwareBacked.toString())
                .put("strongBoxBacked", installationKey.strongBoxBacked.toString()))
            .put("attestation", JSONObject()
                .put("provider", "play_integrity")
                .put("token", attestationToken)
                .put("keyId", installationKey.keyId)
                .put("applicationId", BuildConfig.APPLICATION_ID))
        val response = request("POST", "/api/v1/edge/enroll", payload)
        requireStatus(response, 201)
        val json = JSONObject(response.body)
        return EnrolledNode(json.getString("id"), json.getString("token"))
    }

    suspend fun heartbeat(token: String, batteryPercent: Double?, onBattery: Boolean): ControlCommand {
        val payload = JSONObject()
            .put("metrics", JSONObject()
                .putOpt("batteryPercent", batteryPercent)
                .put("onBattery", onBattery))
            .put("capabilities", JSONObject()
                .put("models", JSONArray().put(configuration.modelName))
                .put("modelProfiles", JSONArray().put(JSONObject()
                    .put("name", configuration.modelName)
                    .put("provider", "llama.cpp")
                    .put("quantization", "gguf")
                    .put("capabilities", JSONArray().put("text"))))
                .put("embeddingModels", JSONArray())
                .put("agentRuntimes", JSONArray().put("single"))
                .put("agentRuntimeProfiles", JSONArray().put("tool_loop_v1"))
                .put("maxConcurrency", 1)
                .put("labels", JSONObject()
                    .put("runtime", "android-service")
                    .put("inference", if (configuration.backend == InferenceBackend.VULKAN) "llama.cpp+vulkan" else "llama.cpp+cpu")))
        val response = request("POST", "/api/v1/workers/heartbeat", payload, token)
        requireStatus(response, 200)
        return control(JSONObject(response.body))
    }

    suspend fun control(token: String): ControlCommand {
        val response = request("GET", "/api/v1/edge/control", null, token)
        requireStatus(response, 200)
        return control(JSONObject(response.body))
    }

    suspend fun lease(token: String): JSONObject? {
        val response = request(
            "POST",
            "/api/v1/workers/lease",
            JSONObject().put("workerVersion", "android/1.6.0"),
            token,
        )
        if (response.status == 204) return null
        requireStatus(response, 200)
        return JSONObject(response.body)
    }

    suspend fun renew(token: String, leaseId: String) {
        requireStatus(request("POST", "/api/v1/leases/${leaseId.pathSegment()}/renew", JSONObject(), token), 204)
    }

    suspend fun complete(token: String, leaseId: String, output: String, durationMs: Long) {
        val payload = JSONObject()
            .put("output", output)
            .put("metrics", JSONObject()
                .put("durationMs", durationMs)
                .put("modelCalls", 1)
                .put("model", configuration.modelName)
                .put("provider", if (configuration.backend == InferenceBackend.VULKAN) "llama.cpp-vulkan" else "llama.cpp-cpu"))
        requireStatus(request("POST", "/api/v1/leases/${leaseId.pathSegment()}/complete", payload, token), 200)
    }

    suspend fun fail(token: String, leaseId: String, message: String) {
        val safe = message.replace(Regex("[\\r\\n\\u0000]"), " ").take(2_000).ifBlank { "Native inference failed" }
        requireStatus(request("POST", "/api/v1/leases/${leaseId.pathSegment()}/fail", JSONObject().put("error", safe), token), 200)
    }

    suspend fun acknowledgeWipe(token: String, generation: Int, localDataDeleted: Boolean) {
        val payload = JSONObject()
            .put("generation", generation)
            .put("credentialsDeleted", true)
            .put("localDataDeleted", localDataDeleted)
        requireStatus(request("POST", "/api/v1/edge/control/wipe-ack", payload, token), 200)
    }

    fun attestationBinding(challenge: EnrollmentChallenge, keyId: String): String = listOf(
        "agat-edge-attestation-v1",
        challenge.id,
        challenge.challenge,
        configuration.nodeName,
        BuildConfig.APPLICATION_ID,
        keyId,
    ).joinToString("\n")

    fun requestHash(binding: String): String = Base64.encodeToString(
        MessageDigest.getInstance("SHA-256").digest(binding.toByteArray(Charsets.UTF_8)),
        Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING,
    )

    private fun control(json: JSONObject): ControlCommand = ControlCommand(
        action = json.getString("action"),
        generation = json.optInt("generation", 0),
        reason = json.optString("reason").takeIf { it.isNotBlank() && it != "null" },
    )

    private suspend fun request(method: String, path: String, body: JSONObject?, bearer: String? = null): HttpResponse =
        withContext(Dispatchers.IO) {
            val connection = URL("$baseUrl$path").openConnection() as HttpURLConnection
            try {
                connection.instanceFollowRedirects = false
                connection.requestMethod = method
                connection.connectTimeout = 10_000
                connection.readTimeout = 70_000
                connection.setRequestProperty("Accept", "application/json")
                connection.setRequestProperty("User-Agent", "agat-edge-android/1.6.0")
                bearer?.let { connection.setRequestProperty("Authorization", "Bearer $it") }
                if (body != null) {
                    connection.setRequestProperty("Content-Type", "application/json")
                    connection.doOutput = true
                    connection.outputStream.use { it.write(body.toString().toByteArray(Charsets.UTF_8)) }
                }
                val status = connection.responseCode
                val stream = if (status in 200..299) connection.inputStream else connection.errorStream
                val responseBody = stream?.use { input ->
                    val output = ByteArrayOutputStream()
                    val buffer = ByteArray(8_192)
                    var total = 0
                    while (true) {
                        val count = input.read(buffer)
                        if (count < 0) break
                        total += count
                        if (total > 1_048_576) throw IOException("Ответ coordinator слишком большой")
                        output.write(buffer, 0, count)
                    }
                    output.toString(Charsets.UTF_8.name())
                }.orEmpty()
                HttpResponse(status, responseBody)
            } finally {
                connection.disconnect()
            }
        }

    private fun requireStatus(response: HttpResponse, expected: Int) {
        if (response.status == expected) return
        val detail = runCatching { JSONObject(response.body).optString("error") }.getOrNull().orEmpty()
        throw IOException("Coordinator HTTP ${response.status}${if (detail.isBlank()) "" else ": ${detail.take(500)}"}")
    }

    private fun String.pathSegment(): String = java.net.URLEncoder.encode(this, Charsets.UTF_8.name()).replace("+", "%20")
    private data class HttpResponse(val status: Int, val body: String)
}
